### PR TITLE
Video 3728 add input field and send button

### PR DESCRIPTION
**Pull request by olipyskoty**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [X] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [VIDEO-3728](https://issues.corp.twilio.com/browse/VIDEO-3728)

### Description

This PR adds an input field and send button to the chat window so the user can type and send messages to other participants. The send button will be highlighted in blue when anything other than whitespace is typed. Users can press `enter` or click the button to send a message, while `enter + shift` creates a new line. Please note that actually sending a message in this PR will throw an error until the new server is complete and merged. We also decided to keep the blue outline around the `<TextareaAutosize />` when it is focused for accessibility reasons. 

**web:**

**mobile:**

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [x] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Included screenshot as PR comment (if needed)
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Re-tested if necessary

### CODE REVIEW

File: src/components/ChatWindow/ChatInput/ChatInput.test.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  it('should call the correct function when send message button is clicked', () => {
+    const wrapper = shallow(<ChatInput conversation={{ sendMessage: mockHandleSendMessage } as any} />);
+    wrapper.find(TextareaAutosize).simulate('change', { target: { value: 'I am a message!!!' } });
+    wrapper
+      .find(SendMessageIcon)
+      .parent()
+      .simulate('click');
+    expect(mockHandleSendMessage).toHaveBeenCalled();
+  });
```
Comment (timmydoza) [inline src/components/ChatWindow/ChatInput/ChatInput.test.tsx:55]: All these tests look good! But I think we could use some additional tests for the `handlereturnKeyPress` function. We'll need to test that it works correctly for the different values of `isMobile`, `event.key`, and `event.shiftKey`.

CircleCI provides a test coverage report for each PR (which can also be generated locally with `npm run test:ci`), which can be used to see if there are any tests that are missing: https://1225-216070925-gh.circle-artifacts.com/0/coverage/lcov-report/src/components/ChatWindow/ChatInput/ChatInput.tsx.html

File: src/components/ChatWindow/ChatInput/ChatInput.test.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  it('should call the correct function when send message button is clicked', () => {
+    const wrapper = shallow(<ChatInput conversation={{ sendMessage: mockHandleSendMessage } as any} />);
+    wrapper.find(TextareaAutosize).simulate('change', { target: { value: 'I am a message!!!' } });
+    wrapper
+      .find(SendMessageIcon)
+      .parent()
+      .simulate('click');
+    expect(mockHandleSendMessage).toHaveBeenCalled();
```
Comment (timmydoza) [inline src/components/ChatWindow/ChatInput/ChatInput.test.tsx:54]: Even better is:

```suggestion
    expect(mockHandleSendMessage).toHaveBeenCalledWith('I am a message!!!');
```

Also, we could have a test to make sure that the whitespace is being trimmed. We could probably just use this same test for that:

`wrapper.find(TextareaAutosize).simulate('change', { target: { value: '  I am a message!!!  \n   ' } });`

File: src/components/ChatWindow/ChatInput/ChatInput.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import { Conversation } from '@twilio/conversations/lib/conversation';
+
+const useStyles = makeStyles({
+  chatInputContainer: {
+    borderTop: '1px solid #e4e7e9',
+    borderBottom: '1px solid #e4e7e9',
+    padding: '1em 1.2em 1em',
+  },
+  textArea: {
+    paddingTop: '0.2em',
```
Comment (timmydoza) [inline src/components/ChatWindow/ChatInput/ChatInput.tsx:15]: I used this for my screenshots in slack:

```
padding: '0.75em 1em`,
marginTop: `0.4em`
```

But feel free to play with it. `padding: '0.75em 1em` makes the padding identical to the padding on the text inputs you see before you enter a room.

**Review by timmydoza — CHANGES_REQUESTED** [review]

File: src/components/ChatWindow/ChatInput/ChatInput.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    background: '#0263E0',
+    cursor: 'pointer',
+  },
+});
+
+interface InputFieldProps {
+  conversation: Conversation;
+}
+
+export default function InputField({ conversation }: InputFieldProps) {
```
[the remark this answers is not in the document]
Comment (olipyskoty) [reply src/components/ChatWindow/ChatInput/ChatInput.tsx:42]: ahh good catch - missed this one 👍

File: src/components/ChatWindow/ChatInput/ChatInput.tsx (hunk 3 above)
Comment (olipyskoty) [reply src/components/ChatWindow/ChatInput/ChatInput.tsx:15]: i think this looks good and i like how it stays consistent with the inputs on the "login" page

**Comment by olipyskoty** [comment]
@timmydoza updating the `<div>` to `<button>` tag in the `<ChatWindowHeader>` now includes the blue outline when tabbing to the `<button>`, or when clicking on it. just fyi - but good for accessibility!

File: src/components/ChatWindow/ChatWindowHeader/ChatWindowHeader.tsx [hunk 5]
```diff
@@ -21,6 +21,8 @@ const useStyles = makeStyles(() =>
     closeChatWindow: {
       cursor: 'pointer',
       display: 'flex',
+      background: 'transparent',
+      border: '0',
```
Comment (timmydoza) [inline src/components/ChatWindow/ChatWindowHeader/ChatWindowHeader.tsx:25]: Nit: I think adding `padding: '0.4em'` helps a little. It makes the blue border square when it is focused. Previously it was a little rectangular.

image

File: src/components/ChatWindow/ChatInput/ChatInput.tsx [hunk 6]
```diff
@@ -70,20 +73,20 @@ export default function InputField({ conversation }: InputFieldProps) {
...
       />
 
-      <div
+      <button
         className={clsx(classes.sendButton, { [classes.activeSendButton]: isValidMessage })}
         onClick={() => handleSendMessage(messageBody)}
       >
         <SendMessageIcon />
-      </div>
+      </button>
```
Comment (timmydoza) [inline src/components/ChatWindow/ChatInput/ChatInput.tsx]: Looks good! But now that this is a button, I think we should use the `disabled` prop when the button isn't supposed to be clickable:

```
      <button className={classes.sendButton} onClick={() => handleSendMessage(messageBody)} disabled={!isValidMessage}>
        <SendMessageIcon />
      </button>
```
When `disabled` is true, it tells the browser that the button isn't supposed to be clickable. As a result, the button won't be tabbable when it's disabled. So I think it goes a little further than just changing the styles so that it doesn't look clickable. 

Also, with this change, we can take advantage of the `:disabled` [pseudo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/:disabled) to apply the right styles when it's disabled. This way we don't have to worry about conditionally applying the class ourselves. So the `sendButton` class can look like:

```
  sendButton: {
    height: '40px',
    width: '40px',
    border: '0',
    borderRadius: '4px',
    float: 'right',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: '1em',
    background: '#0263E0',
    cursor: 'pointer',
    '&:disabled': {
      background: 'none',
      cursor: 'default',
    },
  },
  ```

The `&` is just a substitute for the classname of the parent object. So `'&:disabled'` gets turned into `.sendButton:disabled`.

Then, the test will have to be updated to look for the `disabled` prop instead of the `activeSendButton` class (which we can get rid of). For example:

```
    expect(
      wrapper
        .find(SendMessageIcon)
        .parent()
        .prop('disabled')
    ).toBe(true);
```

**Review by timmydoza — CHANGES_REQUESTED** [review]

File: src/components/ChatWindow/ChatInput/ChatInput.tsx (hunk 6 above)
Comment (olipyskoty) [reply src/components/ChatWindow/ChatInput/ChatInput.tsx]: good point! this looks nicer too

**Review by timmydoza — APPROVED** [review]
